src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPImprovedTextMateTypedHandler.java

```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.formatting;
+
+import com.intellij.codeInsight.editorActions.TypedHandlerDelegate;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiFile;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Improves on-type handling of unbalanced nested brace/bracket/paren pairs in TextMate files.
+ */
+public class LSPImprovedTextMateTypedHandler extends TypedHandlerDelegate {
+
+    @Override
+    @NotNull
+    public Result charTyped(char charTyped,
+                            @NotNull Project project,
+                            @NotNull Editor editor,
+                            @NotNull PsiFile file) {
+        /* TODO: Uncomment once LSP4IJ no longer supports 2023.2
+        if (LSPIJEditorUtils.isSupportedTextMateFile(file) &&
```

Comment:
This is the TypeScript change that's commented out until 2023.2 support is dropped. Hopefully that will happen relatively soon and, once it does, I'll be happy to enable all of this TypeScript-specific functionality that's been committed commented out.